### PR TITLE
BUG: Fix reduction UDFs over ungrouped, bounded windows on Pandas backend

### PR DESCRIPTION
**Pull request by timothydijamco**
## Overview
Currently, trying to apply a reduction UDF over an ungrouped, bounded window on the Pandas backend will hit an error.

This is because the execution rule for reduction UDFs (over an ungrouped aggregation context) **doesn't handle the case that the aggregation context could be a bounded window** (assumes it is unbounded).

This PR adds logic to handle this case.

### Edit: Incidental improvements
This PR also ended up making some improvements to aggregation contexts (`ibis/pandas/aggcontext.py`) and how they are used to execute reduction/analytic UDFs (`ibis/pandas/udf.py`) to simplify things and increase consistency:
1. Removed branching in [`execute_udaf_node_no_groupby`](https://github.com/ibis-project/ibis/pull/2395/files#diff-de9deaf50a2dbf3ee6c6ae3d25b38138R224) (now simply calling `aggcontext.agg` regardless of the specific type of `AggregationContext`)
2. Because of (1), we are using `aggcontext.agg` in more cases. This surfaced an error, which I fixed by **no longer** passing `kwargs` to `aggcontext.agg` (in [`execute_udaf_node_no_groupby`](https://github.com/ibis-project/ibis/pull/2395/files#diff-de9deaf50a2dbf3ee6c6ae3d25b38138R224) and  [`execute_udaf_node_groupby`](https://github.com/ibis-project/ibis/pull/2395/files#diff-de9deaf50a2dbf3ee6c6ae3d25b38138R235)).
This is because reduction/analytic UDFs [already have `kwargs` applied when they were created](https://github.com/ibis-project/ibis/blob/master/ibis/udf/vectorized.py#L41). Trying to pass `kwargs` to the functions again leads to an error (~unexpected keyword argument) in some cases.
3. No longer pass duplicate columns to `aggcontext.agg`.
In other words, always do something like this:
`aggcontext.agg(args[0], func, *args[1:], **kwargs)`
rather than this:
`aggcontext.agg(args[0], func, *args, **kwargs)`

### Code
```
import pandas as pd
import ibis
from ibis.udf.vectorized import reduction

client = ibis.pandas.connect({'test_table': df})

table = client.table('test_table')

@reduction(
    input_type=['double'],
    output_type='double'
)
def my_mean_udf(col):
    return col.mean()

# Ungrouped, bounded window
win = ibis.window(
    following=0,
    order_by=table.id
)

expr = table.mutate(
    val=my_mean_udf(table.value_col).over(win)
)

result = expr.execute()

print(result.to_markdown())
```

#### Before
```
...

~/ibis/ibis/pandas/execution/window.py in execute_window_op(op, data, window, scope, timecontext, aggcontext, clients, **kwargs)
    356         **kwargs,
    357     )
--> 358     series = post_process(result, data, ordering_keys, grouping_keys)
    359     assert len(data) == len(
    360         series

~/ibis/ibis/pandas/execution/window.py in _post_process_order_by(series, parent, order_by, group_by)
     67     if len(names) > 1:
     68         series = series.reorder_levels(names)
---> 69     series = series.iloc[index.argsort(kind='mergesort')]
     70     return series
     71 

AttributeError: 'numpy.float64' object has no attribute 'iloc'
```

#### After
|    |   id |   value_col |     val |
|---:|-----:|------------:|--------:|
|  0 |    0 |           4 | 4       |
|  1 |    1 |           1 | 2.5     |
|  2 |    2 |           2 | 2.33333 |
|  3 |    3 |           3 | 2.5     |

No error, and the result is correct

## Testing
I added a new test for ungrouped, bounded windows in `test_window.py`.

#### Readability changes
I also renamed the tests in `test_window.py` to make it more clear which _variant_ of windows each test is handling. This is what the list of tests in `test_window.py` now look like:
- `test_grouped_bounded_expanding_window`
- `test_ungrouped_bounded_expanding_window` **(new in this PR)**
- `test_grouped_bounded_following_window`
- `test_grouped_bounded_preceding_windows`
- `test_grouped_unbounded_window`
- `test_ungrouped_unbounded_window`

I'm open to any suggestions to try to simplify this, since `test_window.py` keeps growing. I've found this difficult to simplify further because there are so many combinations of windows and operations over windows (several dimensions: grouped/ungrouped, bounded/unbounded, ordered/unordered, reduction/analytic, udf/non-udf). To make things worse, specific permutations may need to be xfailed on certain backends, so there's a minimum amount explicitness required for test parameters (to let us xfail very specific permutations).

### CODE REVIEW

File: ibis/tests/all/test_window.py [hunk 1]
```diff
@@ -224,8 +226,55 @@ def test_window(backend, alltypes, df, con, result_fn, expected_fn):
...
+        ),
+        param(
+            # Disabled on PySpark and Spark backends becuase in pyspark<3.0.0,
+            # Pandas UDFs are only supported on unbounded windows
+            lambda t, win: mean_udf(t.double_col).over(win),
+            lambda df: (df.double_col.expanding().mean()),
+            id='mean_udf',
+            marks=[
+                pytest.mark.udf,
+                pytest.mark.skip_backends([PySpark, Spark]),
```
Comment (timothydijamco) [inline ibis/tests/all/test_window.py:245]: We technically could check the version of `pyspark` here and only skip if `pyspark<3.0.0`, but following in the same vein as #2364, I am avoiding adding package version checks for tests in `tests/all`. This is because `pyspark` might not even be installed here, depending on which CI job is running. I feel a version check would be more complex than its worth

**Review by jreback — APPROVED** [review]
lgtm

File: ibis/pandas/udf.py [hunk 2]
```diff
@@ -216,30 +216,47 @@ def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
     input_type = op.input_type
     nargs = len(input_type)
 
-    # An execution rule for a simple aggregate node
+    # An execution rule for ungrouped aggregations. Handles:
+    #   1) AggregationNode with no grouping
+    #   2) Some aggregation over a window with no grouping
```
Comment (icexelloss) [inline ibis/pandas/udf.py:221]: What do you mean by "Some aggregation"?

Comment (icexelloss) [reply ibis/pandas/udf.py:221]: Can you please explain exactly which aggregation over a window with no grouping is covered by this?

File: ibis/pandas/udf.py [hunk 3]
```diff
@@ -216,30 +216,47 @@ def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
...
     @execute_node.register(type(op), *(itertools.repeat(pd.Series, nargs)))
     def execute_udaf_node(op, *args, **kwargs):
```
Comment (icexelloss) [inline ibis/pandas/udf.py:223]: Perhaps update the name of this to `execute_udaf_node_no_grouping` to clearer

File: ibis/pandas/udf.py [hunk 4]
```diff
@@ -216,30 +216,47 @@ def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
...
-    # An execution rule for a grouped aggregation node. This
-    # includes aggregates applied over a window.
+        func = op.func
+
+        if isinstance(aggcontext, Window):
+            # If aggcontext is a Window, then we must be aggregating over
+            # a bounded window, so we need aggcontext.agg to handle the
+            # rolling/expanding window logic.
+            return aggcontext.agg(args[0], func, *args, **kwargs)
+        else:
```
Comment (icexelloss) [inline ibis/pandas/udf.py:234]: Can we assert exactly the type of aggcontext here? e.g

```
if isinstance(aggcontext, Window):

elif isinstance(aggcontext, ???):

else:
    raise AssertionError(f"Unexpected aggcontext of type {type{asggcontext}")

```

File: ibis/tests/all/test_window.py (hunk 1 above)
Comment (icexelloss) [reply ibis/tests/all/test_window.py:245]: So this works for pyspark 3 but we are still skipping it?

File: ibis/pandas/udf.py (hunk 3 above)
Comment (icexelloss) [reply ibis/pandas/udf.py:223]: or `execute_udaf_node_ungrouped` or that's the naming conversion you are using

**Review by icexelloss — CHANGES_REQUESTED** [review]
Reviewed one around

File: ibis/tests/all/test_window.py (hunk 1 above)
Comment (icexelloss) [reply ibis/tests/all/test_window.py:245]: Oh potential way to do this is to do a try import at the top:

```
try:
    import pyspark
    _pyspark_version = ... # find out spark version
except ImportError:
     _pyspark_version = None
```

Comment (icexelloss) [reply ibis/tests/all/test_window.py:245]: FYI this is how spark does it:
https://github.com/apache/spark/blob/master/python/pyspark/sql/tests/test_pandas_udf.py#L24

We could have perhaps a `ibis.tests.utils` to encapsulate this

Comment (icexelloss) [reply ibis/tests/all/test_window.py:245]: I think this might be adding unnecessary complexity. So let's leave comment here at least and perhaps address as follow up.

File: ibis/pandas/udf.py (hunk 2 above)
Comment (timothydijamco) [reply ibis/pandas/udf.py]: I updated these comments. They were actually inaccurate--these execute rules don't actually handle the Aggregation node. They handle analytic and reduction UDFs.

File: ibis/pandas/udf.py (hunk 3 above)
Comment (timothydijamco) [reply ibis/pandas/udf.py]: Done

File: ibis/pandas/udf.py (hunk 4 above)
Comment (timothydijamco) [reply ibis/pandas/udf.py]: Done

File: ibis/tests/all/test_window.py (hunk 1 above)
Comment (timothydijamco) [reply ibis/tests/all/test_window.py:245]: Sounds good. I have a comment already about that parameter being disabled due to `pyspark<3.0.0` but let me know if you think I should expand the comment

File: ibis/pandas/udf.py (hunk 2 above)
Comment (timothydijamco) [reply ibis/pandas/udf.py]: Apparently it _does_ handle Aggregation node as well! Let met update these comments again.

File: ibis/tests/all/test_window.py (hunk 1 above)
Comment (icexelloss) [reply ibis/tests/all/test_window.py:245]: No the current comment is good.

**Review by icexelloss — APPROVED** [review]
LGTM. +1

File: ibis/tests/all/test_window.py [hunk 5]
```diff
@@ -429,16 +481,17 @@ def test_unbounded_window_ungrouped(
         )
 
     # Define a window that is
-    # 1) Not grouped
+    # 1) Unrouped
```
Comment (icexelloss) [inline ibis/tests/all/test_window.py:484]: This is a typo

**Review by icexelloss — CHANGES_REQUESTED** [review]

**Comment by icexelloss** [comment]
Sorry revoked my approved. There is a small typo otherwise LGTM.

**Comment by timothydijamco** [comment]
I updated the comments to be more accurate and also a little more concise. I also updated when an `AssertionLogic` is thrown to be more accurate. These changes are mostly because I realized that the code I am changing in `ibis/pandas/udf.py` **is** also used by the `Aggregation` node (not just windows), so I've made the comments and logic reflect that now

I also fixed the typo in `test_window.py`

**Comment by timothydijamco** [comment]
CI is failing--looks like there is a use case of creating a custom aggcontext ([link](https://github.com/ibis-project/ibis/blob/816f52187e2111990c6fb7930280fa578534c6aa/ibis/pandas/tests/execution/test_window.py#L674)). Let me look into how to handle this

**Comment by timothydijamco** [comment]
OK, to handle custom `AggregationContexts`, I removed the `AssertionError` that I was raising when the type of the `AggregationContext` is unknown. In place of raising an error, I am now just calling the custom `AggregationContext`'s `agg` method.

#### Note:
For the most part, this is exactly how the execution rules handled custom `AggregationContexts` before this PR.

The only difference is that **before this PR**, when handling a **grouped aggregation**, the execution rule would **pre-process** the data. Below is what the execution rule did **before this PR** in the case that the `AggregationContext` type is `Summarize`, `Transform`, or a custom `AggregationContext`:
https://github.com/ibis-project/ibis/blob/899804c2a9a5fc0e721369a5952a09c5191cdd8b/ibis/pandas/udf.py#L242-L252

In this PR, I'm changing this so that custom `AggregationContexts` don't get this pre-processing. This is what the execution rule for custom `AggregationContexts` looks like in this PR:
https://github.com/ibis-project/ibis/blob/7239cd640e0552e5d5168ec06e2b0e6285a06c3e/ibis/pandas/udf.py#L283-L286

I believe this is more proper. I don't think we should pre-process any data passed to custom `AggregationContexts`, because custom `AggregationContexts` can/should be written to do any processing it needs

**Comment by jreback** [comment]
@icexelloss if you'd have a  look / approve when ready. ignore the 1 failing part of the CI for now.

File: ibis/pandas/udf.py [hunk 6]
```diff
@@ -216,30 +216,56 @@ def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
...
-        return op.func(*args)
+    def execute_udaf_node_no_groupby(op, *args, **kwargs):
+        aggcontext = kwargs.pop('aggcontext', None)
+        assert aggcontext is not None, 'aggcontext is None'
+
+        func = op.func
 
-    # An execution rule for a grouped aggregation node. This
-    # includes aggregates applied over a window.
+        if isinstance(aggcontext, Summarize) or isinstance(
```
Comment (icexelloss) [inline ibis/pandas/udf.py:228]: isinstance(aggcontext, (A, B))

File: ibis/pandas/udf.py [hunk 7]
```diff
@@ -216,30 +216,56 @@ def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
...
         assert aggcontext is not None, 'aggcontext is None'
 
         func = op.func
 
-        if isinstance(aggcontext, Window):
-            # Call the func differently for Window because of
-            # the custom rolling logic.
-            result = aggcontext.agg(args[0], func, *args, **kwargs)
-        else:
+        if isinstance(aggcontext, Summarize) or isinstance(
```
Comment (icexelloss) [inline ibis/pandas/udf.py:255]: Same here

File: ibis/pandas/udf.py [hunk 8]
```diff
@@ -249,8 +275,14 @@ def aggregator(first, *rest, **kwargs):
...
                 return func(first, *map(next, rest))
 
-            result = aggcontext.agg(args[0], aggregator, *iters, **kwargs)
-
-        return result
+            return aggcontext.agg(args[0], aggregator, *iters, **kwargs)
+        elif isinstance(aggcontext, Window):
+            # We must be aggregating over a bounded window, so we need
+            # aggcontext.agg to handle the rolling/expanding window logic.
+            return aggcontext.agg(args[0], func, *args, **kwargs)
```
Comment (icexelloss) [inline ibis/pandas/udf.py:282]: Can you look into if you can pass *args[1:]

Comment (timothydijamco) [reply ibis/pandas/udf.py]: Done!

I had to make a change to `window_agg_udf` to make the `Window` `AggregationContext` handle this properly

**Comment by timothydijamco** [comment]
Changes:

1. Removed branching in [`execute_udaf_node_no_groupby`](https://github.com/ibis-project/ibis/pull/2395/files#diff-de9deaf50a2dbf3ee6c6ae3d25b38138R224) (now simply calling `aggcontext.agg` regardless of the specific type of `AggregationContext`)
2. Because of #1, I had to make a fix to never pass `kwargs` to `aggcontext.agg` in [`execute_udaf_node_no_groupby`](https://github.com/ibis-project/ibis/pull/2395/files#diff-de9deaf50a2dbf3ee6c6ae3d25b38138R224) and  [`execute_udaf_node_groupby`](https://github.com/ibis-project/ibis/pull/2395/files#diff-de9deaf50a2dbf3ee6c6ae3d25b38138R235).
This is because reduction/analytic UDFs [already have `kwargs` applied when they were created](https://github.com/ibis-project/ibis/blob/master/ibis/udf/vectorized.py#L41). Trying to pass `kwargs` to the functions again leads to an error (~unexpected keyword argument) in some cases.
3. No longer pass duplicate columns to `aggcontext.agg`.
In other words, always do something like this:
`aggcontext.agg(args[0], func, *args[1:], **kwargs)`
rather than this:
`aggcontext.agg(args[0], func, *args, **kwargs)`

File: ibis/pandas/aggcontext.py [hunk 9]
```diff
@@ -400,12 +400,11 @@ def window_agg_udf(
...
 
-    # If there is no args, then the UDF only takes a single
-    # input which is defined by grouped_data
-    # This is a complication due to the lack of standard
-    # way to pass multiple input pd.Series/SeriesGroupBy
-    # to AggregationContext.agg()
-    inputs = args if len(args) > 0 else [grouped_data]
+    # The first input column is in grouped_data, but there may
+    # be additional input columns in args.
+    inputs = [grouped_data]
```
Comment (icexelloss) [inline ibis/pandas/aggcontext.py:405]:
```
inputs = [grouped_data] + args
```
Does this work?

File: ibis/pandas/udf.py [hunk 10]
```diff
@@ -216,41 +216,58 @@ def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
...
 
-    # An execution rule for a simple aggregate node
+    # An execution rule to handle analytic and reduction UDFs over
+    # 1) an ungrouped window,
+    # 2) an ungrouped Aggregate node, or
+    # 3) an ungrouped custom aggregation context
     @execute_node.register(type(op), *(itertools.repeat(pd.Series, nargs)))
-    def execute_udaf_node(op, *args, **kwargs):
-        return op.func(*args)
+    def execute_udaf_node_no_groupby(op, *args, **kwargs):
```
Comment (icexelloss) [inline ibis/pandas/udf.py:224]:
```
def execute_udaf_node_no_groupby(op, *args, aggcontext, **kwargs):
```
Does this work instead of poping out `aggcontext` from kwargs?

File: ibis/pandas/aggcontext.py (hunk 9 above)
Comment (icexelloss) [reply ibis/pandas/aggcontext.py:405]: Sorry, I meant
```
inputs = (grouped_data,) + args
```

**Comment by icexelloss** [comment]
@timothydijamco Can you also add https://github.com/ibis-project/ibis/pull/2395#issuecomment-697781351 to the overview section?

File: ibis/pandas/aggcontext.py (hunk 9 above)
Comment (timothydijamco) [reply ibis/pandas/aggcontext.py]: Yep! this works. I've changed it now

File: ibis/pandas/udf.py (hunk 10 above)
Comment (timothydijamco) [reply ibis/pandas/udf.py]: This does work as well. I've made this change

File: ibis/pandas/udf.py (hunk 4 above)
Comment (timothydijamco) [reply ibis/pandas/udf.py]: (Ended up removing this `AssertionError` in a later commit so that we continue to support custom `AggregationContexts`)

**Review by icexelloss — APPROVED** [review]
LGTM. +1

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @timothydijamco
